common/tasks/interleaved_weighted_round_robin.go

```diff
@@ -279,38 +366,101 @@ func (s *InterleavedWeightedRoundRobinScheduler[T, K]) dispatchTasksWithWeight()
 			s.Unlock()
 		}
 
-		weightedChannels := s.channels()
-		s.doDispatchTasksWithWeight(weightedChannels)
+		iwrrChannels := s.channels()
+		enableRateLimiter := s.isRateLimiterEnabled()
+		s.doDispatchTasksWithWeight(iwrrChannels, enableRateLimiter)
+
+		if !enableRateLimiter {
+			continue LoopDispatch
+		}
+
+		// rate limiter enabled
+		// all channels = throttled channels + not throttled but has more task + not throttled and no more task
+		// - If there's channel that's not throttled but has more task, need to trigger next round
+		//   of dispatch immediately.
+		// - Otherwise all channels = throttled channels + not throttled and no more task
+		//   then as long as there's throttled channel, need to set a timer to try dispatch later
+
+		numThrottled := 0
+		for _, channel := range iwrrChannels.channels {
+			if channel.throttled {
+				numThrottled++
+				continue
+			}
+			if channel.moreTasks {
+				// there's channel that is not throttled and may have more tasks
+				// start a new round of dispatch immediately
+				continue LoopDispatch
+			}
+		}
+
+		if numThrottled != 0 {
+			s.setupDispatchTimer()
+		}
+
+		return
 	}
 }
 
 func (s *InterleavedWeightedRoundRobinScheduler[T, K]) doDispatchTasksWithWeight(
-	channels WeightedChannels[T],
+	iwrrChannels iwrrChannels[T, K],
+	enableRateLimiter bool,
 ) {
+	rateLimiter := quotas.NoopRequestRateLimiter
+	if enableRateLimiter {
+		rateLimiter = s.rateLimiter
+		for _, channel := range iwrrChannels.channels {
+			channel.throttled = false
+			channel.moreTasks = false
+		}
+	}
+
+	numFlattenedChannels := len(iwrrChannels.flattenedChannels)
+	startIdx := rand.Intn(numFlattenedChannels)
 	numTasks := int64(0)
 LoopDispatch:
-	for _, channel := range channels {
+	for i := 0; i != numFlattenedChannels; i++ {
```

Comment:
I added a randomization here. Otherwise I need there will be a starvation problem if rate limiter is low and we always using the same order for all the channels.